README.md
# spring-security-adfs-saml2 [![Build Status](https://travis-ci.org/choonchernlim/spring-security-adfs-saml2.svg?branch=master)](https://travis-ci.org/choonchernlim/spring-security-adfs-saml2)

Spring Security module for service provider (Sp) to authenticate against identity provider's (IdP) ADFS using SAML2 protocol.

How this module is configured:-

* `HTTP-Redirect` binding for sending SAML messages to IdP.
* Handles Sp servers doing SSL termination.
* Default authentication method is user/password using IdP's form login page. 
* Default signature algorithm is `SHA256withRSA`.
* Default digest algorithm is `SHA-256`.

Tested against Sp's environments:-

* Local Tomcat server without SSL termination.
* Azure Tomcat server with SSL termination.

Tested against IdP's environments:-

* ADFS 2.0 - Windows Server 2008 R2.
* ADFS 2.1 - Windows Server 2012.

## Maven Dependency

```xml
<dependency>
  <groupId>com.github.choonchernlim</groupId>
  <artifactId>spring-security-adfs-saml2</artifactId>
  <version>0.9.0</version>
</dependency>
```

## Prerequisites

* Java 8.
* Both Sp and IdP must use HTTPS protocol.
* Java’s default keysize is limited to 128-bit key due to US export laws and a few countries’ import laws. So, Java Cryptography Extension (JCE) Unlimited Strength Jurisdiction Policy Files must be installed to allow larger key size, such as 256-bit key.
* Keystore contains the following:- 
    * (REQUIRED) Sp's public/private keys - to generate digital signature before sending SAML messages to IdP.
    * (OPTIONAL) IdP's public certificate - to verify IdP's SAML messages to prevent man-in-the-middle attack. This certificate can also be stored under JDK's cacerts.
* To generate Sp's public/private keys:-

```
keytool -genkeypair \
 -v \
 -keystore /path/to/keystore.jks \
 -storepass mystorepass \
 -alias myapp \
 -dname 'CN=[COMMON-NAME], OU=[ORGANIZATION-UNIT], O=[ORGANIZATION-NAME], L=[CITY-NAME], ST=[STATE-NAME], C=[COUNTRY]' \
 -keypass mykeypass \
 -keyalg RSA \
 -keysize 2048 \
 -sigalg SHA256withRSA
```

* To import IdP's public certificate into keystore:-

```
keytool -importcert \
  -file idp-adfs-server.crt \
  -keystore /path/to/keystore.jks \
  -alias idp-adfs-server \
  -storepass mystorepass
```

## Usage 

### Simplest Configuration

If you are configuring for one IDP server, the easiest approach is to hardcode all the SAML config in the `@Configuration` file.

```java
// Create a Java-based Spring configuration that extends SAMLWebSecurityConfigurerAdapter.
@Configuration
@EnableWebSecurity
class AppSecurityConfig extends SAMLWebSecurityConfigurerAdapter {

    // See `SAMLConfigBean Properties` section below for more info. 
    @Override
    protected SAMLConfigBean samlConfigBean() {
        return new SAMLConfigBeanBuilder()
                .withIdpServerName("idp-server")
                .withSpServerName("sp-server")
                .withSpContextPath("/app")
                .withKeystoreResource(new DefaultResourceLoader().getResource("classpath:keystore.jks"))
                .withKeystorePassword("storepass")
                .withKeystoreAlias("alias")
                .withKeystorePrivateKeyPassword("keypass")
                .withSuccessLoginDefaultUrl("/")
                .withSuccessLogoutUrl("/goodbye")
                .withStoreCsrfTokenInCookie(true)
                .build();
    }

    // This configuration is not needed if your signature algorithm is SHA256withRSA and 
    // digest algorithm is SHA-256. However, if you are using different algorithm(s), then
    // add this bean with the correct algorithms.
    @Bean
    public static SAMLBootstrap samlBootstrap() {
        return new DefaultSAMLBootstrap("RSA",
                                        SignatureConstants.ALGO_ID_SIGNATURE_RSA_SHA512,
                                        SignatureConstants.ALGO_ID_DIGEST_SHA512);
    }

    // call `samlizedConfig(http)` first to decorate `http` with SAML configuration
    // before configuring app specific HTTP security
    @Override
    protected void configure(final HttpSecurity http) throws Exception {
        samlizedConfig(http)
                .authorizeRequests().antMatchers("/admin/**").hasRole("ADMIN")
                .anyRequest().authenticated();
    }

    // call `samlizedConfig(web)` first to decorate `web` with SAML configuration 
    // before configuring app specific web security
    @Override
    public void configure(final WebSecurity web) throws Exception {
        samlizedConfig(web).ignoring().antMatchers("/resources/**");
    }
}
```

### Customizing SSL Verification

By default, the keystore file serves 2 purposes:-

* Acts as a keystore, containing app's public/private key.
* Acts as a truststore, containing IdP's certificate with public key.

If the keystore does not contain IdP's certificate, the SSL verification will fail with the following error when attempting to retrieve IdP's metadata:-

```
PKIX path construction failed for untrusted credential: [subjectName='CN=idp.server.com,OU=IDP,C=US']: unable to find valid certification path to requested target
I/O exception (javax.net.ssl.SSLPeerUnverifiedException) caught when processing request: SSL peer failed hostname validation for name: null
Error retrieving metadata from https://idp.server.com/federationmetadata/2007-06/federationmetadata.xml
```

If you store the IdP's certificate under JDK's truststore (ie: cacerts) and you want the SSL verification to rely on that file, do this:- 

```java
@Configuration
@EnableWebSecurity
class AppSecurityConfig extends SAMLWebSecurityConfigurerAdapter {

    @Override
    protected SAMLConfigBean samlConfigBean() {
        return new SAMLConfigBeanBuilder()
                // ... other configurations
                .withUseJdkCacertsForSslVerification(true)
                .build();
    }

    ...
}
```

### Environment Properties Driven Configuration

If you don't want to use `@Profile` to configure environment-specific security, you may pass the configuration values through environment properties.

To prevent lifecycle loading or circular dependency issues, instead of autowiring `Environment` into the concrete class, use the given autowired `applicationContext` to get hold of the Spring bean.

```java
@Configuration
@EnableWebSecurity
class AppSecurityConfig extends SAMLWebSecurityConfigurerAdapter {

    @Override
    protected SAMLConfigBean samlConfigBean() {
        final Environment env = applicationContext.getBean(Environment.class);
        
        return new SAMLConfigBeanBuilder()
                .withIdpServerName(env.getProperty("idpServerName"))
                .withSpServerName(env.getProperty("spServerName"))
                .withSpContextPath(env.getProperty("spContextPath"))
                .withKeystoreResource(new DefaultResourceLoader().getResource(env.getProperty("keystoreResource")))
                .withKeystorePassword(env.getProperty("keystorePassword"))
                .withKeystoreAlias(env.getProperty("keystoreAlias"))
                .withKeystorePrivateKeyPassword(env.getProperty("keystorePrivateKeyPassword"))
                .withSuccessLoginDefaultUrl(env.getProperty("successLoginDefaultUrl"))
                .withSuccessLogoutUrl(env.getProperty("successLogoutUrl"))
                .withStoreCsrfTokenInCookie(env.getProperty("storeCsrfTokenInCookie"))
                .build();
    }

    ...
}
```

### Database Driven Configuration

You may also configure `SAMLConfigBean` by retrieving the configuration values from database.

Let's assume you have the following Spring JPA repository:-

```java
public interface SecurityConfigRepository extends JpaRepository<SecurityConfigEntity, Long> {
    SecurityConfigEntity findByEnvironment(String environment);
}
```

To prevent lifecycle loading or circular dependency issues, instead of autowiring `SecurityConfigRepository` into the concrete class, use the given autowired `applicationContext` to get hold of the Spring repository bean.

```java
@Configuration
@EnableWebSecurity
class AppSecurityConfig extends SAMLWebSecurityConfigurerAdapter {

    @Override
    protected SAMLConfigBean samlConfigBean() {
        final SecurityConfigRepository repository = applicationContext.getBean(SecurityConfigRepository.class);
        final SecurityConfigEntity entity = repository.findByEnvironment("dev");
        
        return new SAMLConfigBeanBuilder()
                .withIdpServerName(entity.getIdpServerName())
                .withSpServerName(entity.getSpServerName())
                .withSpContextPath(entity.getSpContextPath())
                .withKeystoreResource(new DefaultResourceLoader().getResource(entity.getKeystoreResource()))
                .withKeystorePassword(entity.getKeystorePassword())
                .withKeystoreAlias(entity.getKeystoreAlias())
                .withKeystorePrivateKeyPassword(entity.getKeystorePrivateKeyPassword())
                .withSuccessLoginDefaultUrl(entity.getSuccessLoginDefaultUrl())
                .withSuccessLogoutUrl(entity.getSuccessLogoutUrl())
                .withStoreCsrfTokenInCookie(entity.getStoreCsrfTokenInCookie())
                .build();
    }

    ...
}
```

### Mocking Security by Hardcoding a Given User for Rapid App Development

```java
@Override
protected void configure(final HttpSecurity http) throws Exception {
    // `CurrentUser` must extend `User`
    final CurrentUser currentUser = new CurrentUser("First name", "Last Name", "ROLE_ADMIN");

    mockSecurity(http, currentUser)
               .authorizeRequests().antMatchers("/admin/**").hasRole("ADMIN")
               .anyRequest().authenticated();
}
```

## SAMLConfigBean Properties
 
`SAMLConfigBean` stores app-specific security configuration.

|Property                        |Required? |Description                                                                                               |
|--------------------------------|----------|----------------------------------------------------------------------------------------------------------|
|idpServerName                   |Yes       |IdP server name. Used for retrieving IdP metadata using HTTPS. If IdP link is `https://idp-server/adfs/ls`, value should be `idp-server`.                                                                                |
|spServerName                    |Yes       |Sp server name. Used for generating correct SAML endpoints in Sp metadata to handle servers doing SSL termination. If Sp link is `https://sp-server:8443/myapp`, value should be `sp-server`.                            |
|spHttpsPort                     |No        |Sp HTTPS port. Used for generating correct SAML endpoints in Sp metadata to handle servers doing SSL termination. If Sp link is `https://sp-server:8443/myapp`, value should be `8443`. <br/><br/>Default is `443`.      |
|spContextPath                   |No        |Sp context path. Used for generating correct SAML endpoints in Sp metadata to handle servers doing SSL termination. If Sp link is `https://sp-server:8443/myapp`, value should be `/myapp`. <br/><br/>Default is `''`.   |
|keystoreResource                |Yes       |App's keystore containing its public/private key and ADFS' certificate with public key.                   |
|keystorePassword                |Yes       |Password to access app's keystore.                                                                        |
|keystoreAlias                   |Yes       |Alias of app's public/private key pair.                                                                   |
|keystorePrivateKeyPassword      |Yes       |Password to access app's private key.                                                                     |
|successLoginDefaultUrl          |Yes       |Where to redirect user on successful login if no saved request is found in the session.                   |
|successLogoutUrl                |Yes       |Where to redirect user on successful logout.                                                              |
|failedLoginDefaultUrl           |No        |Where to redirect user on failed login. This value is set to null, which returns 401 error code on failed login. But, in theory, this will never be used because IdP will handled the failed login on IdP login page.<br/><br/>Default is `''`, which return 401 error code.|
|storeCsrfTokenInCookie          |No        |Whether to store CSRF token in cookie named `XSRF-TOKEN` and expecting CSRF token to be set using header named `X-XSRF-TOKEN` to cater single-page app using frameworks like React and AngularJS. <br/><br/>Default is `false`.             |
|samlUserDetailsService          |No        |For configuring user details and authorities. When set, `userDetails` will be set as `principal`.<br/><br/>Default is `null`. |
|authnContexts                   |No        |Determine what authentication methods to use. To use the order of authentication methods defined by IdP, set as empty set. To enable Windows Integrated Auth (WIA), use `CustomAuthnContext.WINDOWS_INTEGRATED_AUTHN_CTX`.<br/><br/>Default is `AuthnContext.PASSWORD_AUTHN_CTX` where IdP login page is displayed to obtain user/password.|
|useJdkCacertsForSslVerification |No        |When performing IdP's SSL verification, find IdP's certs under JDK's cacerts instead of app's keystore file.<br/><br/>Default is `false`.|


## Important SAML Endpoints

There are several SAML processing endpoints, but these are the ones you probably care:-

|Endpoint              |Description                                                                                                                             |
|----------------------|----------------------------------------------------------------------------------------------------------------------------------------|
|`/saml/login`         |Initiates login process between Sp and IdP. Upon successful login, user will be redirected to `SAMLConfigBean.successLoginDefaultUrl`.  |
|`/saml/logout`        |Initiates logout process between Sp and IdP. Upon successful logout, user will be redirected to `SAMLConfigBean.successLogoutUrl`.      |
|`/saml/metadata`      |Returns Sp metadata. IdP may need this link to register Sp on ADFS.                                                                     |

## Relevant Links

* [Travis CI Reports](https://travis-ci.org/choonchernlim/spring-security-adfs-saml2)
* [Maven Site](https://choonchernlim.github.io/spring-security-adfs-saml2/project-info.html)

Learn about my pains and lessons learned while building this module.

* [Replacing SHA-1 with SHA-256 on Signature and Digest Algorithms](http://myshittycode.com/2016/02/23/spring-security-saml-replacing-sha-1-with-sha-256-on-signature-and-digest-algorithms/)
* [Handling IdP’s Public Certificate When Loading Metadata Over HTTPS](http://myshittycode.com/2016/02/19/spring-security-saml-handling-idps-public-certificate-when-loading-metadata-over-https/)
* [Configuring Binding for Sending SAML Messages to IdP](http://myshittycode.com/2016/02/18/spring-security-saml-configuring-binding-for-sending-saml-messages-to-idp/)
* [Java + SAML: Illegal Key Size](http://myshittycode.com/2016/02/18/java-saml-illegal-key-size/)



CHANGELOG.md
# Change Log

## 0.9.0 - 2019-01-09

* FEATURE - Added `SAMLConfigBean.useJdkCacertsForSslVerification` flag to allow SSL verifications to be performed by using JDK's cacerts instead of app's keystore file.

## 0.8.0 - 2018-07-11

* Moved from Java 7 to Java 8.
* Dependencies update... organized POM.

## 0.7.1 - 2018-05-04

* BUG - `CsrfHeaderFilter` creates multiple cookies with same name but different path due to possible empty context path, which then uses current request's path. This may cause client side to read the wrong cookie when retrieving the CSRF token.

## 0.7.0 - 2017-11-28

* Dropped autowired `Environment` from `SAMLWebSecurityConfigurerAdapter` and replaced with `ApplicationContext` to allow concrete class to access any Spring beans instead of just `Environment` to configure the security. This will also prevent any lifecycle or circular dependency problems when trying to autowire beans in concrete class.
* Replaced `@PostContruct` with `@Bean` for `SAMLWebSecurityConfigurerAdapter.socketFactoryInitialization()`.

## 0.6.0 - 2016-07-18

* Helper class `JndiBackedKeystoreService` to retrieve keystore info from JNDI value with following format: `jks-path,alias,storepass,keypass`

## 0.5.0 - 2016-07-13

* If `samlConfigBean.storeCsrfTokenInCookie` is `true`, then store CSRF token in cookie.
* Decoupled `WebSSOProfileOptions` from `SAMLEntryPoint` to allow user to override `SAMLEntryPoint` easily.
* Dependency updates.

```                  
[INFO]   cglib:cglib-nodep ..................................... 3.2.2 -> 3.2.4
[INFO]   org.codehaus.groovy:groovy-all ........................ 2.4.6 -> 2.4.7
[INFO]   org.spockframework:spock-core ...
[INFO]                                    1.0-groovy-2.4 -> 1.1-groovy-2.4-rc-1
[INFO]   org.springframework:spring-test ....... 4.2.6.RELEASE -> 4.3.1.RELEASE
[INFO]   org.springframework.security:spring-security-config ...
[INFO]                                           4.1.0.RELEASE -> 4.1.1.RELEASE
[INFO]   org.springframework.security:spring-security-core ...
[INFO]                                           4.1.0.RELEASE -> 4.1.1.RELEASE
[INFO]   org.springframework.security:spring-security-web ...
[INFO]                                           4.1.0.RELEASE -> 4.1.1.RELEASE
```                  

## 0.4.0 - 2016-06-05

* If `samlConfigBean.samlUserDetailsService` is provided, then set `samlAuthenticationProvider.forcePrincipalAsString` to `false` so that `principal` represents the `userDetails` object.
* Ability to mock security to bypass authentication against ADFS during rapid app development. To use this, `samlConfigBean.samlUserDetailsService` must be set.
* Dependency, parent and plugins updates.

```                  
com.github.choonchernlim:build-reports ................ 0.2.4 -> 0.3.2
com.google.guava:guava-testlib .......................... 18.0 -> 19.0
junit:junit ............................................. 4.11 -> 4.12
org.codehaus.groovy:groovy-all .............. 2.4.3 -> 2.4.6
org.springframework.security:spring-security-config ...
                                        4.0.3.RELEASE -> 4.1.0.RELEASE
org.springframework.security:spring-security-core ...
                                        4.0.3.RELEASE -> 4.1.0.RELEASE
org.springframework.security:spring-security-web ...
                                        4.0.3.RELEASE -> 4.1.0.RELEASE
org.springframework.security.extensions:spring-security-saml2-core ...
                                        1.0.1.RELEASE -> 1.0.2.RELEASE
maven-compiler-plugin ................................... 3.3 -> 3.5.1
```                       

## 0.3.3 - 2016-04-13                       
* Inject Spring environment to get access to project properties file. ([#1](https://github.com/choonchernlim/spring-security-adfs-saml2/pull/1))

## 0.3.2 - 2016-03-14

* Used `SAMLContextProviderLB` instead of `SAMLContextProviderImpl` to handle servers doing SSL termination.
* Dropped `SAMLConfigBean.spMetadataBaseUrl`.
* Renamed `SAMLConfigBean.adfsHostName` to `SAMLConfigBean.idpHostName`.
* Added `SAMLConfigBean.spServerName`.
* Added `SAMLConfigBean.spHttpsPort`.
* Added `SAMLConfigBean.spContextPath`.

## 0.3.1 - 2016-03-10

* Added `SAMLConfigBean.spMetadataBaseUrl` to manually specify the Sp's metadata base URL to handle situations where servers do SSL termination (HTTPS -> HTTP).
* Configured metadata generator to use user defined Sp's metadata base URL when generating SAML endpoints URLs.

## 0.2.2 - 2016-03-08

* Fixed casing typo from `SAMLConfigBean.keyStoreResource` to `SAMLConfigBean.keystoreResource`.

## 0.2.1 - 2016-03-07

* Added `SAMLConfigBean.keystorePrivateKeyPassword` to add password for private key.
* Kept storepass and keypass separate.
* Excluded `xml-apis` from dependency because it's known to cause problems in WAS.

## 0.2.0 - 2016-03-02

* Options to allow different authentication method. Default is user/password using IdP's form login page.
* `CustomAuthnContext.WINDOWS_INTEGRATED_AUTHN_CTX` to allow Windows Integrated Authentication.

## 0.1.0 - 2016-02-28

* Initial.


src/main/java/com/github/choonchernlim/security/adfs/saml2/SAMLConfigBean.java
package com.github.choonchernlim.security.adfs.saml2;

import static com.github.choonchernlim.betterPreconditions.preconditions.PreconditionFactory.expect;
import com.google.common.collect.ImmutableSet;
import net.karneim.pojobuilder.GeneratePojoBuilder;
import org.opensaml.saml2.core.AuthnContext;
import org.springframework.core.io.Resource;
import org.springframework.security.saml.userdetails.SAMLUserDetailsService;

import java.util.Optional;
import java.util.Set;

/**
 * This class contains all properties that can be configured by Sp using the provided builder class.
 */
public final class SAMLConfigBean {

    /**
     * (REQUIRED) IdP's server name.
     */
    private final String idpServerName;

    /**
     * (REQUIRED) Sp's server name.
     */
    private final String spServerName;

    /**
     * (OPTIONAL) Sp's HTTPS port.
     * <p/>
     * Default is 443.
     */
    private final Integer spHttpsPort;

    /**
     * (OPTIONAL) Sp's context path.
     * <p/>
     * Default is "".
     */
    private final String spContextPath;

    /**
     * (REQUIRED) Keystore containing app's public/private key and ADFS' certificate with public key.
     */
    private final Resource keystoreResource;

    /**
     * (REQUIRED) Keystore alias.
     */
    private final String keystoreAlias;

    /**
     * (REQUIRED) Keystore password.
     */
    private final String keystorePassword;

    /**
     * (REQUIRED) Keystore private key password.
     */
    private final String keystorePrivateKeyPassword;

    /**
     * (REQUIRED) Where to redirect user on successful login if no saved request is found in the session.
     */
    private final String successLoginDefaultUrl;

    /**
     * (REQUIRED) Where to redirect user on successful logout.
     */
    private final String successLogoutUrl;

    /**
     * Where to redirect user on failed login. This value is set to null, which returns
     * 401 error code on failed login. But, in theory, this will never be used because
     * IdP will handled the failed login on IdP login page.
     * <p/>
     * Default is blank.
     */
    private final String failedLoginDefaultUrl;

    /**
     * For configuring user details and authorities.
     * <p/>
     * Default is null.
     */
    private final SAMLUserDetailsService samlUserDetailsService;

    /**
     * Whether to store CSRF token in cookie.
     * </p>
     * Default is false.
     */
    private final Boolean storeCsrfTokenInCookie;

    /**
     * Determine what authentication methods to use.
     * <p/>
     * To use the order of authentication methods defined by IdP, set as empty set.
     * <p/>
     * To enable Windows Integrated Auth (WIA) cross browsers and OSes, use `CustomAuthnContext.WINDOWS_INTEGRATED_AUTHN_CTX`.
     * <p/>
     * Default is user/password authentication where IdP login page is displayed.
     */
    private final Set<String> authnContexts;

    /**
     * Whether to rely on JDK's cacerts for SSL verification or not.
     * <p/>
     * By default, the provided keystore contains ADFS' certificate(s) to perform SSL verification.
     * </p>
     * Default is false.
     */
    private final Boolean useJdkCacertsForSslVerification;

    @GeneratePojoBuilder
    SAMLConfigBean(final String idpServerName,
                   final String spServerName,
                   final Integer spHttpsPort,
                   final String spContextPath,
                   final Resource keystoreResource,
                   final String keystoreAlias,
                   final String keystorePassword,
                   final String keystorePrivateKeyPassword,
                   final String successLoginDefaultUrl,
                   final String successLogoutUrl,
                   final String failedLoginDefaultUrl,
                   final Boolean storeCsrfTokenInCookie,
                   final SAMLUserDetailsService samlUserDetailsService,
                   final Set<String> authnContexts,
                   final Boolean useJdkCacertsForSslVerification) {

        //@formatter:off
        this.idpServerName = expect(idpServerName, "IdP server name").not().toBeBlank().check();

        this.spServerName = expect(spServerName, "Sp server name").not().toBeBlank().check();
        this.spHttpsPort = Optional.ofNullable(spHttpsPort).orElse(443);
        this.spContextPath = Optional.ofNullable(spContextPath).orElse("");

        this.keystoreResource = (Resource) expect(keystoreResource, "Key store").not().toBeNull().check();
        this.keystoreAlias = expect(keystoreAlias, "Keystore alias").not().toBeBlank().check();
        this.keystorePassword = expect(keystorePassword, "Keystore password").not().toBeBlank().check();
        this.keystorePrivateKeyPassword = expect(keystorePrivateKeyPassword, "Keystore private key password").not().toBeBlank().check();

        this.successLoginDefaultUrl = expect(successLoginDefaultUrl, "Success login URL").not().toBeBlank().check();
        this.successLogoutUrl = expect(successLogoutUrl, "Success logout URL").not().toBeBlank().check();
        this.failedLoginDefaultUrl = Optional.ofNullable(failedLoginDefaultUrl).orElse("");

        this.storeCsrfTokenInCookie = Optional.ofNullable(storeCsrfTokenInCookie).orElse( false);
        this.samlUserDetailsService = samlUserDetailsService;

        this.authnContexts = Optional.ofNullable(authnContexts).orElse(ImmutableSet.of(AuthnContext.PASSWORD_AUTHN_CTX));
        this.useJdkCacertsForSslVerification = Optional.ofNullable(useJdkCacertsForSslVerification).orElse(false);
        //@formatter:on
    }

    public String getIdpServerName() {
        return idpServerName;
    }

    public String getSpServerName() {
        return spServerName;
    }

    public Integer getSpHttpsPort() {
        return spHttpsPort;
    }

    public String getSpContextPath() {
        return spContextPath;
    }

    public Resource getKeystoreResource() {
        return keystoreResource;
    }

    public String getKeystoreAlias() {
        return keystoreAlias;
    }

    public String getKeystorePassword() {
        return keystorePassword;
    }

    public String getKeystorePrivateKeyPassword() {
        return keystorePrivateKeyPassword;
    }

    public String getSuccessLoginDefaultUrl() {
        return successLoginDefaultUrl;
    }

    public String getSuccessLogoutUrl() {
        return successLogoutUrl;
    }

    public String getFailedLoginDefaultUrl() {
        return failedLoginDefaultUrl;
    }

    public Boolean getStoreCsrfTokenInCookie() {
        return storeCsrfTokenInCookie;
    }

    public SAMLUserDetailsService getSamlUserDetailsService() {
        return samlUserDetailsService;
    }

    public Set<String> getAuthnContexts() {
        return authnContexts;
    }

    public Boolean getUseJdkCacertsForSslVerification() {
        return useJdkCacertsForSslVerification;
    }
}
